### ISSUE TITLE
`clippy::manual-dangling-ptr` : doesn't have a size known at compile-time

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#16459]
Using the following flags

`--force-warn clippy::manual-dangling-ptr`

this code:

```rust
#![feature(extern_types)]

use std::ops::Deref;
use std::pin::Pin;

unsafe extern "C" {
    type Extern;
}

impl Deref for &mut Extern {
    type Target = i32;
    fn deref(&self) -> &i32 { &1 }
}

fn make_extern() -> &'static mut Extern {
    // Suppose that this allocates and leaks an Extern on the heap
    unsafe { &mut *(1 as *mut Extern) }
}

fn main() {
    let r: &'static mut Extern = make_extern();
    // This is allowed because &mut Extern dereferences to i32,
    // which is Unpin
    let mut i = 0;
}
```

caused the following diagnostics:

```
    Checking _fecd5337d43314f926793127002f908f2c716a6d v0.1.0 (/tmp/icemaker_global_tempdir.GToAtTsoC3qp/icemaker_clippyfix_tempdir.iheKAq168pQw/_fecd5337d43314f926793127002f908f2c716a6d)
warning: manual creation of a dangling pointer
  --> src/main.rs:17:20
   |
17 |     unsafe { &mut *(1 as *mut Extern) }
   |                    ^^^^^^^^^^^^^^^^^^ help: use: `std::ptr::dangling_mut::<Extern>()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_dangling_ptr
   = note: requested on the command line with `--force-warn clippy::manual-dangling-ptr`

warning: `_fecd5337d43314f926793127002f908f2c716a6d` (bin "_fecd5337d43314f926793127002f908f2c716a6d") generated 1 warning (run `cargo clippy --fix --bin "_fecd5337d43314f926793127002f908f2c716a6d" -p _fecd5337d43314f926793127002f908f2c716a6d` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.23s

```

However after applying these diagnostics, the resulting code:6

```rust
#![feature(extern_types)]

use std::ops::Deref;
use std::pin::Pin;

unsafe extern "C" {
    type Extern;
}

impl Deref for &mut Extern {
    type Target = i32;
    fn deref(&self) -> &i32 { &1 }
}

fn make_extern() -> &'static mut Extern {
    // Suppose that this allocates and leaks an Extern on the heap
    unsafe { &mut *std::ptr::dangling_mut::<Extern>() }
}

fn main() {
    let r: &'static mut Extern = make_extern();
    // This is allowed because &mut Extern dereferences to i32,
    // which is Unpin
    let mut i = 0;
}
```
no longer compiled:

```
    Checking _fecd5337d43314f926793127002f908f2c716a6d v0.1.0 (/tmp/icemaker_global_tempdir.GToAtTsoC3qp/icemaker_clippyfix_tempdir.iheKAq168pQw/_fecd5337d43314f926793127002f908f2c716a6d)
error[E0277]: the size for values of type `Extern` cannot be known at compilation time
   --> src/main.rs:17:45
    |
 17 |     unsafe { &mut *std::ptr::dangling_mut::<Extern>() }
    |                                             ^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `std::marker::Sized` is not implemented for `Extern`
note: required by an implicit `Sized` bound in `std::ptr::dangling_mut`
   --> /home/matthias/.rustup/toolchains/master/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:946:27
    |
946 | pub const fn dangling_mut<T>() -> *mut T {
    |                           ^ required by the implicit `Sized` requirement on this type parameter in `dangling_mut`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `_fecd5337d43314f926793127002f908f2c716a6d` (bin "_fecd5337d43314f926793127002f908f2c716a6d") due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `_fecd5337d43314f926793127002f908f2c716a6d` (bin "_fecd5337d43314f926793127002f908f2c716a6d" test) due to 1 previous error

```

Version:
```
rustc 1.95.0-nightly (75963ce79 2026-01-25)
binary: rustc
commit-hash: 75963ce795666bc1f961e5d60060809809f6bc68
commit-date: 2026-01-25
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 21.1.8

```
